pkg/tcpip/network/ipv4/ipv4.go

```diff
@@ -283,7 +283,9 @@ func (e *endpoint) WriteHeaderIncludedPacket(r *stack.Route, payload buffer.Vect
 		return nil
 	}
 
-	hdr := buffer.NewPrependableFromView(payload.ToView())
+	// If we want to send the packet to a link-layer,
+	// we have to reserve space for an Ethernet header.
+	hdr := buffer.NewPrependableFromView(payload.ToView(), int(e.MaxHeaderLength()))
```

Comment:
I think you should be using:
e.linkEP.MaxHeaderLength()

e.MaxHeaderLength() includes the IP header.

Reply:
fixed